[MRG + 2] FIX Be robust to non re-entrant/ non deterministic cv.split calls

raghavrv opened the pull request:
Fixes #6726 
- To make consistent splits for all parameter settings when the cv iterators are non re-entrant (thanks to Joel for the word!) and non deterministic for successive `split` calls, we call split once and store it as a list that will be iterated multiple times.
- Similarly in `_CVIterableWrapper`, we store the list rather than the cv object and iterate on it... (NOTE that this would make previously (intentionally) non deterministic cv iterables, deterministic with consistent output for successive `split` calls. Do we want that to happen?)

raghavrv committed:
FIX Be robust to non re-entrant/ non deterministic cv.split calls

raghavrv commented:
Ping @jnothman @amueller for review...

raghavrv commented:
Also what kind of tests should I add here? I can make a mock splitter which will iterate only once and pass it as cv to grid search. That will fail in master but will pass in this branch. Any other tests?

jnothman commented:
passing in a generator expression for `cv` would be a sufficient non-regression test.

Without tests, surely this should be WIP?

raghavrv commented:
> Without tests, surely this should be WIP?

Indeed. Sorry for marking it MRG

raghavrv committed:
TST Ensure custom one time iterators are accepted as cv

raghavrv commented:
Done! Reviews please ;)

raghavrv committed:
FIX/TST pop the time keys

raghavrv committed:
FLAKE8

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -129,6 +128,7 @@ def cross_val_score(estimator, X, y=None, groups=None, scoring=None, cv=None,
     X, y, groups = indexable(X, y, groups)
 
     cv = check_cv(cv, y, classifier=is_classifier(estimator))
+    cv_iter = cv.split(X, y, groups)
```
why this change?

raghavrv replied:
Just as a sign of best practice denoting that we enlist the indices and store them as soon as we check the `cv` param...

amueller replied:
I'm not sure I get the point. ``cv_iter`` is a generator, right?

raghavrv replied:
argh sorry I somehow kept thinking I did `list(cv.split(...)`...

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -384,6 +384,7 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
     X, y, groups = indexable(X, y, groups)
 
     cv = check_cv(cv, y, classifier=is_classifier(estimator))
+    cv_iter = cv.split(X, y, groups)
```
why?

raghavrv replied:
<Same comment as above>

Should I leave it as before?

amueller commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1154,3 +1154,33 @@ def test_search_train_scores_set_to_false():
+                'mean_fit_time', 'std_fit_time'):
+        gs.cv_results_.pop(key)
+        gs2.cv_results_.pop(key)
+    np.testing.assert_equal(gs.cv_results_, gs2.cv_results_)
```
why not scikit learn testing?

raghavrv replied:
The numpy's `testing.assert_equal` is to check equivalence of nested dicts. We don't have an equivalent AFAIK...

amueller commented:
you're not testing the consistency of folds in ``GridSearchCV`` or ``learning_curve`` or ``validation_curve``.

raghavrv committed:
TST consistency of folds across params in GSCV

amueller commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1154,3 +1154,52 @@ def test_search_train_scores_set_to_false():
+        def __init__(self, n_samples=100):
+            self.indices = KFold(n_splits=5).split(np.ones(n_samples))
+
+        def split(self, X=None, y=None, groups=None):
```
maybe add a comment that says that this ``split`` can only be called once.

amueller commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1154,3 +1154,52 @@ def test_search_train_scores_set_to_false():
+                            _pop_time_keys(gs2.cv_results_))
+
+    # Check consistency of folds across the parameters
+    gs = GridSearchCV(LinearSVC(random_state=0),
```
Did this test fail before? Maybe also add a one-line comment explaining what's happening.

raghavrv replied:
Yes it fails on master...

raghavrv committed:
Enlist the cv folds

raghavrv committed:
Move CustomSplitter to the top. Add explanatory comment

raghavrv committed:
Add test to check cv splits consistency in validation_curve

raghavrv committed:
FIX/Improve tests for cv consistency

raghavrv committed:
FIX/TST cv consistency tests in grid search

raghavrv committed:
Arggh. This won't test consistency of splits. The scores not dependent on splits

raghavrv committed:
enlist

raghavrv commented:
@amueller Have fixed your comments. I'm not able to find a way to test the consistency of cv folds in `learning_curve`, the learning curve seems to filter duplicate values for train sizes. Otherwise I could have given two similar values and test their scores as a proof of consistency in splits...

Other than that, I have addressed all your comments, added explanatory comments in tests + all tests pass. One more review please?

cc: @jnothman too if you find time... :)

amueller commented on sklearn/model_selection/_split.py:
```diff
@@ -1467,7 +1467,7 @@ def get_n_splits(self, X=None, y=None, groups=None):
     """Wrapper class for old style cv objects and iterables."""
     def __init__(self, cv):
-        self.cv = cv
+        self.cv = list(cv)
```
I think this change is fine. This is only used internally, right? All sklearn functions worked with a single-run iterator before, and now they will do that again after this PR. So there are no behavior changes, right?
If someone used ``check_cv`` they will get something different now, which is not entirely great, but this is a bug-fix on 0.18 so we have to change behavior. We never promised our wrapper would be one-iteration if the original cv was.

raghavrv replied:
(Just to note that I've added the 2nd test just to ensure this change has effect and to fail if reverted later...)

jnothman replied:
I suppose infinite CV strategies are unlikely. Expensive ones are a bit more likely, but I think this approach is okay.

raghavrv replied:
Do we revert this change too? @jnothman @amueller?

amueller commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -61,6 +61,7 @@
 from sklearn.datasets import make_multilabel_classification
 
 from sklearn.model_selection.tests.test_split import MockClassifier
```
I don't think tests should import from other tests.

raghavrv replied:
There are quite a few `MockClassifier`'s that are specific to the tests and the module. I'm not sure there is much benefit in grouping them into `utils.testing`...

raghavrv replied:
> Well if they are specific to a test, then I think they should stay there.

They are specific to 2 sets of test in 2 different test files, but not anywhere else except the `model_selection` tests, hence my disinclination to move them to `utils.testing` :)

Maybe you want them moved to `sklearn.model_selection.tests._test_setup`?

amueller commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -573,6 +577,18 @@ def test_learning_curve():
+            cv=CustomSplitter(n_splits=n_splits, n_samples=30),
+            train_sizes=np.linspace(0.1, 1.0, 10))
+    if len(w) > 0:
+        raise RuntimeError("Unexpected warning: %r" % w[0].message)
```
AssertionError?

amueller replied:
Or just do an ``assert`` instead of an ``if``?

raghavrv committed:
TST Add test for _CVIterableWrapper to check for reusability

jnothman commented:
Elsewhere we have ./sklearn/cluster/tests/common.py

On 20 October 2016 at 01:10, Raghav RV <notifications@github.com> wrote:

> *@raghavrv* commented on this pull request.
> ------------------------------
>
> In sklearn/model_selection/tests/test_validation.py
> <https://github.com/scikit-learn/scikit-learn/pull/7660>:
>
> > @@ -61,6 +61,7 @@
>  from sklearn.datasets import make_multilabel_classification
>
>  from sklearn.model_selection.tests.test_split import MockClassifier
>
> Well if they are specific to a test, then I think they should stay there.
>
> They are specific to 2 sets of test in 2 different test files, but not
> anywhere else except the model_selection tests, hence my disinclination
> to move them to utils.testing :)
>
> Maybe you want them moved to sklearn.model_selection.tests._test_setup?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7660>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6zAtCX1DnzgNqZ-zMHJ7hicL9514ks5q1iTvgaJpZM4KVa_v>
> .
>

raghavrv commented:
> Elsewhere we have ./sklearn/cluster/tests/common.py

Thanks for checking. I'll move them to `model_selection/tests/common.py`...

amueller commented:
needs rebase

raghavrv commented:
Done :)

amueller commented:
LGTM if test pass

raghavrv commented:
@jnothman A second review please?

jnothman commented:
I'll try to handle this today; if I do not, I won't get to it for about a week.

jnothman commented on sklearn/model_selection/tests/common.py:
```diff
@@ -0,0 +1,23 @@
+    def __init__(self, n_splits=4, n_samples=99):
+        self.n_splits = n_splits
+        self.n_samples = n_samples
+        self.indices = KFold(n_splits=n_splits).split(np.ones(n_samples))
```
if split returns a list, then the premise of this test is false. use `iter()` to make sure it's not reentrant.

jnothman commented on sklearn/model_selection/tests/common.py:
```diff
@@ -0,0 +1,23 @@
+from sklearn.model_selection import KFold
+
+
+class CustomSplitter:
```
Call it `OneTimeSplitter`

jnothman reviewed (changes requested):
At a skim, this otherwise lgtm

raghavrv committed:
CustomSplitter --> OneTimeSplitter; use iter to ensure one time splitting

raghavrv commented:
Done. Please merge if you are happy :)

raghavrv committed:
Flake8

TomDLT commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1154,3 +1156,58 @@ def test_search_train_scores_set_to_false():
+    # once if ``cv.split`` is called once per param setting in GridSearchCV.fit
+    # the 2nd and 3rd parameter will not be evaluated as no train/test indices
+    # will be generated for the 2nd and subsequent cv.split calls.
+    # This is a check to make cv.split is not called once per param
```
... to make [sure] cv.split is called ...?

raghavrv replied:
Thx!!

raghavrv committed:
TYPO

amueller commented:
I think waiting for @jnothman to come back would be ok ;) I'll try to make sure I have reviewed everything for 0.18.1 till then (unless @TomDLT or someone else want's to give a :+1: of course)

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -908,6 +844,14 @@ def test_cv_iterable_wrapper():
+    kf_iter_wrapped = check_cv(kf_iter)
+    # Since the wrapped iterable is enlisted and stored,
+    # split can be called any number of times to produce
+    # consistent results
```
Only true because it is not random. Please note this. Indeed, it may be good to test that randomization produces different splits in repeated calls to `split`.

raghavrv replied:
In this test, we are wrapping the iter, not the splitter and hence even with randomization, it will produce similar splits only... `check_cv(KFold(n_splits=5, shuffle=True).split(X, y))`

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -764,6 +843,40 @@ def test_validation_curve():
+                                                  n_samples=n_samples))
+    # The OneTimeSplitter is a non-re-entrant cv splitter. Unless, the
+    # `split` is called for each parameter, the following should produce
+    # identical results for param setting 1 param setting 2 as both have
```
"for param setting 1 *and* param setting 2"

raghavrv replied:
Fixed :)

jnothman reviewed (changes requested):
very minor. Otherwise good to merge.

raghavrv committed:
COSMIT

raghavrv committed:
TST Check if shuffle=True produces different splits

raghavrv commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -848,9 +848,23 @@ def test_cv_iterable_wrapper():
+        pass
+    else:
+        raise AssertionError("After randomization, wrapped kfold (with "
+                             "shuffle) did not produce differing results.")
```
Are these the tests you wanted @jnothman @amueller?

amueller replied:
semantically sort of. But that's a really weird way to write it. How about
```python
assert_false(list(kf_randomized_iter.split(X, y)) == list(kf_randomized_iter.split(X,y))
```

raghavrv replied:
Nevermind... I just used np.any + np.array to make it more easy to follow! Thx!

raghavrv commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -908,6 +844,28 @@ def test_cv_iterable_wrapper():
+    # If the splits are randomized, successive calls to split yields different
+    # results
+    kf_randomized_iter = KFold(n_splits=5, shuffle=True).split(X, y)
+    kf_randomized_iter_wrapped = check_cv(kf_iter)
```
`kf_iter` --> `kf_randomized_iter`

raghavrv committed:
TYPO

raghavrv committed:
Use np.any+np.array to assert inequality of splits

amueller commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -908,6 +844,22 @@ def test_cv_iterable_wrapper():
+    kf_randomized_iter_wrapped = check_cv(kf_randomized_iter)
+    assert_array_equal(list(kf_randomized_iter_wrapped.split(X, y)),
+                       list(kf_randomized_iter_wrapped.split(X, y)))
+    assert_true(np.any(np.array(list(kf_iter_wrapped.split(X, y))) !=
```
do you need ``list`` here? anyhow, :+1:

raghavrv replied:
Yea `split(X, y)` is a generator...

GaelVaroquaux replied:
Well, generators can be iterated upon several times without being consumed. It's iterators that cannot (a generator is an object that creates an iterator upon iteration. The "yield" keyword makes generators).

jnothman replied:
```
>>> def a():
...     yield 5
...
>>> x = a()
>>> type(x)
<class 'generator'>
>>> list(x)
[5]
>>> list(x)
[]
```
I think you mean that generator functions can be called many times??

amueller commented:
@jnothman I think this is good now?

raghavrv commented:
Can I haz merge?

jnothman commented:
LGTM

jnothman commented:
I do think there is room for a bit more documentation of the `split` interface, saying that multiple calls to `split` may result in different splits, unless `random_state` is an integer.

amueller commented:
@raghavrv wanna do the doc addition @jnothman suggested?

raghavrv commented:
Sure. It's on my 0.18.1 todo list! I'll send a PR tomorrow...

GaelVaroquaux commented on sklearn/model_selection/_search.py:
```diff
@@ -550,6 +550,7 @@ def _fit(self, X, y, groups, parameter_iterable):
         base_estimator = clone(self.estimator)
         pre_dispatch = self.pre_dispatch
 
+        cv_iter = list(cv.split(X, y, groups))
```
Would a copy of the cv object have done the trick?

jnothman commented:
cv.split can return a generator or other iterable that can be exhausted. A
tee might suffice, but not a copy.

On 2 November 2016 at 18:38, Gael Varoquaux <notifications@github.com>
wrote:

> *@GaelVaroquaux* commented on this pull request.
> ------------------------------
>
> In sklearn/model_selection/_search.py
> <https://github.com/scikit-learn/scikit-learn/pull/7660#pullrequestreview-6761208>
> :
>
> > @@ -550,6 +550,7 @@ def _fit(self, X, y, groups, parameter_iterable):
>          base_estimator = clone(self.estimator)
>          pre_dispatch = self.pre_dispatch
>
> +        cv_iter = list(cv.split(X, y, groups))
>
> Would a copy of the cv object have done the trick?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7660#pullrequestreview-6761208>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69r_h51FL_SJQsdm6pBxp7JH3sl6ks5q6D4AgaJpZM4KVa_v>
> .
>

GaelVaroquaux commented on sklearn/model_selection/_validation.py:
```diff
@@ -961,6 +960,7 @@ def validation_curve(estimator, X, y, param_name, param_range, groups=None,
     X, y, groups = indexable(X, y, groups)
 
     cv = check_cv(cv, y, classifier=is_classifier(estimator))
+    cv_iter = list(cv.split(X, y, groups))
```
Here too: wouldn't a copy of the cv object have done the trick.

raghavrv replied:
I think here the list can be safely removed without side effects as the iteration is only once and is based on splits...

GaelVaroquaux commented:
@jnothman : I don't understand: if there a copy, the copy is exhausted, but not the original.

Maybe I wasn't clear: I was thinking of doing the copy in the list comprehension, so that the list comprehension exhausts the copy, and not the original.

lesteve commented:
> Maybe I wasn't clear: I was thinking of doing the copy in the list comprehension

How would you do the copy? `copy.(deep)copy` won't fly because generators can not be pickled. `itertools.tee` as pointed out by @jnothman could work.

GaelVaroquaux commented:
> How would you do the copy? copy.(deep)copy won't fly because generators can not be pickled.

The following works:

<pre>
def f():
    for i in range(10):
        yield i


import copy

g = copy.copy(f)
</pre>


But not the following:
<pre>
j = (i for i in range(10))

h = copy.copy(j)
</pre>

So I guess that it should work with the CV objects, no?

> itertools.tee as pointed out by @jnothman could work.

That might also be a good idea.

jnothman commented:
You can copy the generator function, but not the generator.

jnothman commented:
`split(..)` evaluates to a generator, not a generator function.

We could call `split` multiple times in gscv, but at the risk that multiple calls return distinct splits, as they currently do if `random_state` is a non-int in any that support `shuffle=True`.

jnothman commented:
i wouldn't usually support materialising an iterator as a list unnecessarily, but I think this is the solution of least surprise to the user, especially as CV splits aren't usually expensive to compute or store relative to the rest of the problem, and their number is necessarily finite and predetermined, at least in gscv/rscv.

amueller commented:
this needs a whatsnew.

GaelVaroquaux commented:
I think that the solutions used in this PR are not correct: it is important to rely on iterators and not lists. Lists are leading to heavy data duplication https://github.com/scikit-learn/scikit-learn/issues/7919

raghavrv commented:
Ahh okay should we resort to [itertools.tee](https://docs.python.org/2/library/itertools.html#itertools.tee) then?

jnothman commented:
itertools.tee effectively stores things. We may just have to prohibit
non-reentrant iterators??

On 22 November 2016 at 10:05, Raghav RV <notifications@github.com> wrote:

> Ahh okay should we resort to itertools.tee
> <https://docs.python.org/2/library/itertools.html#itertools.tee> then?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7660#issuecomment-262096161>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz690iKqNOWD-N12f1bWVIQNKtdVXfks5rAiPAgaJpZM4KVa_v>
> .
>

raghavrv commented:
I'm +1 for that. That would simplify a few things + discourage hacked up `cv`s that do not conform to our splitter `API` (Though I am not sure if reentrancy is kind of implied by our API)...

jnothman commented:
I think this still comes back to things like randomization between calls to `split()`. I'll try think about this more later.

:(

amueller commented:
@GaelVaroquaux how do you ensure that an generator returns identical iterators when called twice without storing it?
So we could give up on guaranteeing consistency across runs using arbitrary cross-validation iterators, and make sure ours are generating consistent sequences?

What do you think should be the semantics of passing ``KFold(shuffle=True)`` to ``GridSearchCV``?
Our previous behavior (pre 0.18 and post 0.18.1) is that all parameters have the same folds.
In fact, iterating of ``KFold(shuffle=True)`` any number of times used to create exactly the same folds. Since 0.18, each iteration over ``KFold(shuffle=True)`` returns different folds.
And that is something we were not really careful enough about :(

I would argue that having it return always the same sequence is somewhat counter-intuitive. What do you think?

GaelVaroquaux commented:
> @GaelVaroquaux how do you ensure that an generator returns identical iterators
> when called twice without storing it?

Well, a generator strictly speak not. But we could ensure that the split
method of the CV object returns the same thing.

> I would argue that having it return always the same sequence is somewhat
> counter-intuitive. What do you think?

In terms of intuition, maybe if you didn't specific the random_state, I
would agree with you.

However, that's not a nice semantic, because it will require the user to
store the list of indices generated in many cases. For instance when
running multiple estimators and wanting to compare the results. It will
be easy to get it wrong. If you look at the bug that was fixed by
@raghavrv, I believe that the code would be wrong without the list if the
iteration was not reproducible.
https://github.com/scikit-learn/scikit-learn/pull/7660/files#diff-676ff6854a280680ed7d91103ef4b0b5R553

The list expands the generator, and hence avoids any irreproducibility
and side effects. But in a sense it defeats the purpose of having
generators. If we consider the extreme case of really large n small p
(say p = 30, and n in the millions), doing 100 iterations of
shuffle-split would create more indices that are stored in the memory
than actual data.

amueller commented:
The semantics of always producing the same result are somewhat nicer than what's there now, and yes, using lists defeats the purpose of iterators.
We can go back to checking the ``random_state`` on ``__init__`` if we agree that "deterministic randomness" is fine. We should be very explicit about that, though.

GaelVaroquaux commented:
> We can go back to checking the random_state on __init__ if we agree that
> "deterministic randomness" is fine. We should be very explicit about that,
> though.

With insight, I think that this is better. The bug and corresponding fix
point it out clearly, IMHO.

raghavrv commented:
> Since 0.18, each iteration over KFold(shuffle=True) returns different folds.

If that is the only concern here we could change the order of iteration [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/model_selection/_search.py#L565). It would paralelize over parameters rather than splits. The problem with that approach is it could be slower when one parameter setting happens to train faster than another, unless joblib is smart to figure that out and let the next job replace that core. (For instance say I iterate over `n_estimators=[1, 10, 100, 2000]` of random forest, then for a 4 core machine the first core could be idle unless joblib has some nice heuristics that I am unaware of..)

jnothman commented:
Joblib has no problem with pooling and taking the next job off the queue afaik, so that's not an issue.

raghavrv commented:
Shall I proceed with the PR. Are you okay with that approach @amueller @GaelVaroquaux?

raghavrv commented:
And for the problem of shuffle in spliltters what do you think of setting a random int as `random_state` if `random_state` is `None`.

And then reusing this guy to regenerate rng at `split`?

```py

# At init
if random_state is None:
    random_state = check_random_state(random_state).randint(0, 100000)

# At split
rng = check_random_state(random_state)
```

Would take of determinism for successive splits and will also ensure the splits are random for multiple initialization but consistent across splits after initialization.

raghavrv commented:
If, after both my changes, all tests pass we are good to go I think...

raghavrv commented:
> we could change the order of iteration here.

Sweeeet now we have to worry about one time iterating parameter samplers / grids ;@

raghavrv commented:
I think we can have #7935, revert all the `list(iter)` changes brought about by this PR and add a note at cv doc stating that multiple iterations are done and hence we expect the cv to not be a non-reentrant one. If they want they can materialize the list externally... WDYT?

raghavrv commented:
> Sweeeet now we have to worry about one time parameter samplers / grid iterators

Wait since all the parameter dicts are anyway generated and stored, it wouldn't actually hurt to materialize that as a list and store them. If my intuition here is correct, doing this will not consume any additional memory...

jnothman commented:
Yes, we store parameter lists anyway.

On 26 November 2016 at 03:16, Raghav RV <notifications@github.com> wrote:

> Sweeeet now we have to worry about one time parameter samplers / grid
> iterators
>
> Wait since all the parameter dicts are anyway generated and stored, it
> wouldn't actually hurt to materialize that as a list and store them. If my
> intuition here is correct, doing this will not consume any additional
> memory...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7660#issuecomment-262989881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_P41GgVCASELpUSfkZ_kZQhP3KMks5rBwn1gaJpZM4KVa_v>
> .
>
